xds/internal/balancer/priority/balancer.go

```diff
@@ -0,0 +1,221 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package priority
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/internal/buffer"
+	"google.golang.org/grpc/internal/grpclog"
+	"google.golang.org/grpc/internal/hierarchy"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/xds/internal/balancer/balancergroup"
+)
+
+const priorityBalancerName = "priority_experimental"
+
+func init() {
+	balancer.Register(&priorityBB{})
+}
+
+type priorityBB struct{}
+
+func (pbb *priorityBB) Build(cc balancer.ClientConn, _ balancer.BuildOptions) balancer.Balancer {
+	return newPriorityBalancer(cc)
+}
+
+func (pbb *priorityBB) Name() string {
+	return priorityBalancerName
+}
+
+type priorityBalancer struct {
+	logger           *grpclog.PrefixLogger
+	cc               balancer.ClientConn
+	bg               *balancergroup.BalancerGroup
+	ctx              context.Context
+	cancel           context.CancelFunc
+	childStateUpdate *buffer.Unbounded
+
+	config *lbConfig
+
+	mu            sync.Mutex
+	childInUse    string
+	priorityInUse int
+	// priorities is a list of child names from higher to lower priority.
+	priorities      []string
+	childToPriority map[string]int
+	// children is a map from child name to sub-balancers.
+	children map[string]*childBalancer
+	// The timer to give a priority 10 seconds to connect. And if the priority
+	// doesn't go into Ready/Failure, start the next priority.
+	//
+	// One timer is enough because there can be at most one priority in init
+	// state.
+	priorityInitTimer *time.Timer
+}
+
+func newPriorityBalancer(cc balancer.ClientConn) *priorityBalancer {
+	b := &priorityBalancer{
+		cc:               cc,
+		childToPriority:  make(map[string]int),
```

Comment:
I'm guessing we don't need to `make` the `childToPriority` map here since it is always being made again in `UpdateClientConnState`.